### PR TITLE
lock protect nullability cache of symbolic regex node

### PR DESCRIPTION
**Pull request by veanes**
Added lock to protect` SymbolicRegexNode._nullabilityCache` that stores conditional nullability of a node for a given context, for thread-safety. This computation is not the common case as it only applies when the node (regex) starts with an anchor and can potentially be _nullable_ (accept the empty string). Initial thought was to special case nullability for context 0 using a field but this is already covered in most common cases when the regex is neither nullable nor can be nullable  that is checked before. The cache could potentially be moved to the builder as a shared cache to avoid the caches in the nodes but would then create bigger probability of thread contention at the builder level.

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharKind.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharKind.cs:36]: OK. I'll add the explanation and replace it with ContextSize = 64. The explanation being that context represents the pair (kind, kind) using bit-shifting where each kind needs 3 bits as there are 5 different kinds (0..4).

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:180]: I'll use the new constant `ContextSize` here too.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:185]: OK. As a follow-up we might need to revisit some other places (thread-safety of `_statearray` and `_delta` in `SymbolicRegexBuilder`).

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:29]: will do.

**Comment by veanes** [comment]
@stephentoub : I took care of the comments and also **simplified the initial test** in `IsNullableFor` in the last commit. Once the checks pass the code should be merged into main. Just wanted to note that I don't have merge permission here yet, which I don't really mind (but I also cannot choose reviewers).

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs:363]:
```suggestion
                return Volatile.Read(ref builder._delta[offset]) ?? matcher.CreateNewTransition(currentState, minterm, offset);
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs:394]:
```suggestion
                    DfaMatchingState<TSetType> p = Volatile.Read(ref builder._delta[offset]) ?? matcher.CreateNewTransition(nextStates, minterm, offset);
```

**Review by stephentoub — APPROVED** [review]

**Pull request merged**
